docs/config.md
# Configuration

This project uses configuration options to set the behavior of its executables.
All executables that use autograder resources use the same configuration infrastructure
and can therefore be configured the same way and with the same options.
See the [Configuration Options section](#configuration-options) for all available options.

Options can be set on the command-line using the `-c`/`--config` flag.
For example:
```
./bin/logs-example --config log.level=debug
```

Options can also be set using environmental variables by prefixing the option keys
with `AUTOGRADER__` and replacing any `.` with `__`.
For example option key `docker.disable` can be set by:
```
AUTOGRADER__DOCKER__DISABLE='true' ./scripts/run_tests.sh
```

## Directories

The primary directory the autograder will use for storing information is referred to as the "work directory",
and is set to `<base dir>/<instance name>`.
Most other paths are configured to be relative to the work directory.

The base directory is set through the `dirs.base` option,
and defaults to `$XDG_DATA_HOME`.

The instance name is a way to configuration a unique name for your autograder instance.
It can be set with the `instance.name` option,
and defaults to `autograder`.

## Loading Options

Configurations will be loaded in the following order (later options override earlier ones):
 0. The command-line options are checked for `BASE_DIR`.
 1. Load options from environmental variables.
 2. Options are loaded from `WORK_DIR/config` (config.json then secrets.json).
 3. Options are loaded from the current working directory (config.json then secrets.json).
 4. Options are loaded from any files specified with `--config-path` (ordered by appearance).
 5. Options are loaded from the command-line (`--config` / `-c`).

The base directory (`dirs.base`) can ONLY be set via the command-line or environmental variables.
This prevents cycles from the base directory changing and loading new options.

## Configuration Options

Below are all configuration options available for the autograder server.
You can also see all available options by looking in the [config/options.go](../internal/config/options.go) file,
or using the `cmd/list-options` executable.
```
go run cmd/list-options/main.go
```

| Key                            | Type    | Default Value  | Description |
|--------------------------------|---------|----------------|-------------|
| `analysis.individual.poolsize` | Integer | 1               | The number of parallel workers per course when computing individual analysis. |
| `analysis.pairwise.poolsize`   | Integer | 1               | The number of parallel workers per course when computing pairwise analysis. |
| `build.keep`                   | Boolean | false           | Keep artifacts/dirs used when building (not building the server itself, but things like assignment images). |
| `db.type`                      | String  | "disk"          | The type of database to use. |
| `db.pg.uri`                    | String  |                 | Connection string to connect to a Postgres Database. Empty if not using Postgres. |
| `dirs.base`                    | String  | [$XDG_DATA_HOME](https://specifications.freedesktop.org/basedir-spec/latest/) | The base dir for autograder to store data. SHOULD NOT be set in config files (to prevent cycles), only on the command-line. |
| `dirs.backup`                  | String  | dirs.base       | Path to where backups are made. Defaults to inside BASE_DIR. |
| `docker.disable`               | Boolean | false           | Disable the use of docker (usually for testing). |
| `docker.output.maxsize`        | Integer | 4096 (4 MB)     | The maximum allowed size (in KB) for stdout and stderr combined. The default is 4096 KB (4 MB). |
| `email.from`                   | String  |                 | From address for emails sent from the autograder. |
| `email.host`                   | String  |                 | SMTP host for emails sent from the autograder. |
| `email.pass`                   | String  |                 | SMTP password for emails sent from the autograder. |
| `email.port`                   | String  |                 | SMTP port for emails sent from the autograder. |
| `email.user`                   | String  |                 | SMTP username for emails sent from the autograder. |
| `email.smtp.idle`              | Integer | 120000 (2 mins) | Consider an SMTP connection idle if no emails are sent for this number of milliseconds. |
| `email.smtp.minperiod`         | Integer | 250             | Allow for at least this amount of time (in milliseconds) between sending emails. |
| `grading.runtime.max`          | Integer | 300 (5 mins)    | The maximum number of seconds a grader can be running for. |
| `http.store`                   | String  |                 | Store HTTP requests made by the server to the specified directory. |
| `instance.name`                | String  | "autograder"    | A name to identify this autograder instance. Should only contain alphanumerics and underscores. |
| `lockmanager.staleduration`    | Integer | 7200 (2 hours)  | Number of seconds a lock can be unused before getting removed. |
| `log.text.level`               | String  | "INFO"          | The default logging level for the text (stderr) logger. |
| `log.backend.level`            | String  | "INFO"          | The default logging level for the backend (database) logger. |
| `tasks.disable`                | Boolean | false           | Disable all scheduled tasks. |
| `tasks.minrest`                | Integer | 300 (5 mins)    | The minimum time (in seconds) between invocations of the same task. A task instance that tries to run too quickly will be skipped. |
| `testing`                      | Boolean | false           | Assume tests are being run, which may alter some operations. |
| `testdata.load`                | Boolean | false           | Load test data when the database opens. |
| `web.http.port`                | Integer | 8080            | The port to serve HTTP traffic on. Standard is 80 (but requires root to use). |
| `web.http.redirect`            | Boolean | false           | Redirect HTTP traffic to HTTPS. Only used if HTTPS is enabled. |
| `web.https.enable`             | Boolean | false           | Enable HTTPS. A certificate and key must be provided. |
| `web.https.port`               | Integer | 8081            | The port to serve HTTPS traffic on. Standard is 443 (but requires root to use). |
| `web.https.cert`               | String  |                 | The path to a PEM encoded SSL certificate file. The certificate file may contain intermediate certificates following the leaf certificate to form a certificate chain. |
| `web.https.key`                | String  |                 | The path to a PEM encoded SSL key file. |
| `web.maxsize`                  | Integer | 2048 (2 MB)     | The maximum allowed file size (in KB) submitted via POST request. The default is 2048 KB (2 MB). |
| `web.static.root`              | String  |                 | The root directory to serve as part of the static portion of the API. Defaults to empty string, which indicates the embedded static directory. |
| `web.static.fallback`          | Boolean | false           | For any unmatched route (potential 404) that does not have an API prefix, try to match it in the static root before giving the final 404. |


internal/config/options.go
package config

var (
	// Base
	NAME = MustNewStringOption("instance.name", "autograder",
		"A name to identify this autograder instance."+
			" Should only contain alphanumerics and underscores.")

	// Directories
	BASE_DIR = MustNewStringOption("dirs.base", GetDefaultBaseDir(),
		"The base dir for autograder to store data."+
			" SHOULD NOT be set in config files (to prevent cycles), only on the command-line or ENV.")
	BACKUP_DIR = MustNewStringOption("dirs.backup", "", "Path to where backups are made. Defaults to inside BASE_DIR.")

	// Debugging / Testing
	KEEP_BUILD_DIRS   = MustNewBoolOption("build.keep", false, "Keep artifacts/dirs used when building (not building the server itself, but things like assignment images).")
	UNIT_TESTING_MODE = MustNewBoolOption("testing", false, "Assume tests are being run, which may alter some operations.")
	LOAD_TEST_DATA    = MustNewBoolOption("testdata.load", false, "Load test data when the database opens.")
	STORE_HTTP        = MustNewStringOption("http.store", "", "Store HTTP requests made by the server to the specified directory.")
	CI_DOCKER_BUILD   = MustNewBoolOption("ci.docker.build", false, "Inform the system that we are running in a Docker build step of a CI. There are some tests that are flaky in this environment.")

	// Logging
	LOG_TEXT_LEVEL    = MustNewStringOption("log.text.level", "INFO", "The default logging level for the text (stderr) logger.")
	LOG_BACKEND_LEVEL = MustNewStringOption("log.backend.level", "INFO", "The default logging level for the backend (database) logger.")

	// Stats
	STATS_SYSTEM_INTERVAL_MS = MustNewIntOption("stats.system.interval", 60*1000, "The number of milliseconds between system stats collection events.")

	// Email
	EMAIL_FROM                 = MustNewStringOption("email.from", "", "From address for emails sent from the autograder.")
	EMAIL_HOST                 = MustNewStringOption("email.host", "", "SMTP host for emails sent from the autograder.")
	EMAIL_PASS                 = MustNewStringOption("email.pass", "", "SMTP password for emails sent from the autograder.")
	EMAIL_PORT                 = MustNewStringOption("email.port", "", "SMTP port for emails sent from the autograder.")
	EMAIL_USER                 = MustNewStringOption("email.user", "", "SMTP username for emails sent from the autograder.")
	EMAIL_SMTP_IDLE_TIMEOUT_MS = MustNewIntOption("email.smtp.idle", 2*60*1000, "Consider an SMTP connection idle if no emails are sent for this number of milliseconds.")
	EMAIL_MIN_PERIOD           = MustNewIntOption("email.smtp.minperiod", 250, "The minimum time (in MS) between sending emails.")

	// Docker
	DOCKER_DISABLE            = MustNewBoolOption("docker.disable", false, "Disable the use of docker (usually for testing).")
	DOCKER_MAX_OUTPUT_SIZE_KB = MustNewIntOption("docker.output.maxsize", 4*1024, "The maximum allowed size (in KB) for stdout and stderr combined. The default is 4096 KB (4 MB).")

	// Grading
	GRADING_RUNTIME_MAX_SECS = MustNewIntOption("grading.runtime.max", 60*5, "The maximum number of seconds a Docker container can be running for.")

	// Tasks
	NO_TASKS             = MustNewBoolOption("tasks.disable", false, "Disable all scheduled tasks.")
	TASK_MAX_WAIT_SECS   = MustNewIntOption("tasks.maxwait", 2*60, "The maximum wait between checking for the next task to run.")
	TASK_MIN_PERIOD_SECS = MustNewIntOption("tasks.minperiod", 10*60, "The minimum period between the runs of the same task.")

	// Server
	WEB_HTTP_PORT        = MustNewIntOption("web.http.port", 8080, "The port to serve HTTP traffic on. Standard is 80 (but requires root to use).")
	WEB_HTTP_REDIRECT    = MustNewBoolOption("web.http.redirect", true, "Redirect HTTP traffic to HTTPS. Only used if HTTPS is enabled.")
	WEB_HTTPS_ENABLE     = MustNewBoolOption("web.https.enable", false, "Enable HTTPS. A certificate and key must be provided.")
	WEB_HTTPS_PORT       = MustNewIntOption("web.https.port", 8081, "The port to serve HTTPS traffic on. Standard is 443 (but requires root to use).")
	WEB_HTTPS_CERT       = MustNewStringOption("web.https.cert", "", "The path to a PEM encoded SSL certificate file. The certificate file may contain intermediate certificates following the leaf certificate to form a certificate chain.")
	WEB_HTTPS_KEY        = MustNewStringOption("web.https.key", "", "The path to a PEM encoded SSL key file.")
	WEB_MAX_FILE_SIZE_KB = MustNewIntOption("web.maxsize", 2*1024, "The maximum allowed file size (in KB) submitted via POST request. The default is 2048 KB (2 MB).")
	WEB_STATIC_ROOT      = MustNewStringOption("web.static.root", "", "The root directory to serve as part of the static portion of the API. Defaults to empty string, which indicates the embedded static directory.")
	WEB_STATIC_FALLBACK  = MustNewBoolOption("web.static.fallback", false, "For any unmatched route (potential 404) that does not have an API prefix, try to match it in the static root before giving the final 404.")

	// Database
	DB_TYPE   = MustNewStringOption("db.type", "disk", "The type of database to use.")
	DB_PG_URI = MustNewStringOption("db.pg.uri", "", "Connection string to connect to a Postgres Database. Empty if not using Postgres.")

	// Job Management
	ANALYSIS_INDIVIDUAL_COURSE_POOL_SIZE = MustNewIntOption("analysis.individual.poolsize", 1, "The number of parallel workers per course when computing individual analysis.")
	ANALYSIS_PAIRWISE_COURSE_POOL_SIZE   = MustNewIntOption("analysis.pairwise.poolsize", 1, "The number of parallel workers per course when computing pairwise analysis.")
	REGRADE_COURSE_POOL_SIZE             = MustNewIntOption("regrade.poolsize", 4, "The number of parallel workers per course when regrading an assignment.")

	STALELOCK_DURATION_SECS = MustNewIntOption("lockmanager.staleduration", 2*60*60, "Number of seconds a lock can be unused before getting removed.")
)
